File: backend/src/datasource/dbdao/ai_generation.rs
#![allow(dead_code)]
use super::dao::DBDao;
use anyhow::Result;
use serde::{Deserialize, Serialize};
use sqlx::Row;

#[derive(Deserialize, Serialize, Debug, Clone)]
pub struct GeneratorRecord {
    pub id: i64,
    pub name: String,
    pub provider: String,
    pub model: String,
    pub api_key: Option<String>,
    pub api_endpoint: Option<String>,
    pub enabled: bool,
    pub created_at: Option<String>,
}

#[derive(Deserialize, Serialize, Debug, Clone)]
pub struct PromptTemplateRecord {
    pub id: i64,
    pub name: String,
    pub description: Option<String>,
    pub template: String,
    pub input_variables: Option<String>,
    pub provider: Option<String>,
    pub created_at: Option<String>,
}

#[derive(Deserialize, Serialize, Debug, Clone)]
pub struct GenerateTaskRecord {
    pub id: i64,
    pub generator_id: i64,
    pub input: String,
    pub output: Option<String>,
    pub status: Option<String>,
    pub created_at: Option<String>,
}

impl DBDao {
    /// Get generator by ID
    pub async fn get_generator(&self, id: i64) -> Result<GeneratorRecord> {
        let row = sqlx::query(
            "SELECT id, name, provider, model, api_key, api_endpoint, enabled, created_at
             FROM ai_generators WHERE id = $1",
        )
        .bind(id)
        .fetch_one(&self.db)
        .await?;

        let generator = GeneratorRecord {
            id: row.get("id"),
            name: row.get("name"),
            provider: row.get("provider"),
            model: row.get("model"),
            api_key: row.get("api_key"),
            api_endpoint: row.get("api_endpoint"),
            enabled: row.get("enabled"),
            created_at: row.get("created_at"),
        };

        Ok(generator)
    }

    /// Search generators with pagination
    pub async fn search_generators(
        &self,
        provider: Option<String>,
        _model: Option<String>,
        _enabled: Option<bool>,
        page: i64,
        limit: i64,
    ) -> Result<Vec<GeneratorRecord>> {
        let offset = (page - 1) * limit;
        let rows = sqlx::query(
            "SELECT id, name, provider, model, api_key, api_endpoint, enabled, created_at
             FROM ai_generators
             WHERE ($1 IS NULL OR provider = $1)
             OFFSET $2 LIMIT $3",
        )
        .bind(provider)
        .bind(offset)
        .bind(limit)
        .fetch_all(&self.db)
        .await?;

        let generators = rows
            .into_iter()
            .map(|row| GeneratorRecord {
                id: row.get("id"),
                name: row.get("name"),
                provider: row.get("provider"),
                model: row.get("model"),
                api_key: row.get("api_key"),
                api_endpoint: row.get("api_endpoint"),
                enabled: row.get("enabled"),
                created_at: row.get("created_at"),
            })
            .collect::<Vec<_>>();

        Ok(generators)
    }

    /// Add generator
    pub async fn add_generator(
        &self,
        name: &str,
        provider: &str,
        model: &str,
        api_key: Option<String>,
        api_endpoint: Option<String>,
        enabled: bool,
    ) -> Result<i64> {
        let row = sqlx::query(
            "INSERT INTO ai_generators (name, provider, model, api_key, api_endpoint, enabled)
             VALUES ($1, $2, $3, $4, $5, $6)
             RETURNING id",
        )
        .bind(name)
        .bind(provider)
        .bind(model)
        .bind(api_key)
        .bind(api_endpoint)
        .bind(enabled)
        .fetch_one(&self.db)
        .await?;

        Ok(row.get("id"))
    }

    /// Update generator
    pub async fn update_generator(
        &self,
        id: i64,
        name: Option<String>,
        provider: Option<String>,
        model: Option<String>,
        api_key: Option<String>,
        api_endpoint: Option<String>,
        enabled: Option<bool>,
    ) -> Result<i64> {
        let row = sqlx::query(
            "UPDATE ai_generators
             SET name = COALESCE($1, name),
                 provider = COALESCE($2, provider),
                 model = COALESCE($3, model),
                 api_key = COALESCE($4, api_key),
                 api_endpoint = COALESCE($5, api_endpoint),
                 enabled = COALESCE($6, enabled)
             WHERE id = $7
             RETURNING id",
        )
        .bind(name)
        .bind(provider)
        .bind(model)
        .bind(api_key)
        .bind(api_endpoint)
        .bind(enabled)
        .bind(id)
        .fetch_one(&self.db)
        .await?;

        Ok(row.get("id"))
    }

    /// Generate content
    pub async fn generate_content(&self, generator_id: i64, input: &str) -> Result<i64> {
        let row = sqlx::query(
            "INSERT INTO ai_generation_tasks (generator_id, input)
             VALUES ($1, $2)
             RETURNING id",
        )
        .bind(generator_id)
        .bind(input)
        .fetch_one(&self.db)
        .await?;

        Ok(row.get("id"))
    }

    /// Get prompt template by ID
    pub async fn get_prompt_template(&self, id: i64) -> Result<PromptTemplateRecord> {
        let row = sqlx::query(
            "SELECT id, name, description, template, input_variables, provider, created_at
             FROM ai_prompt_templates WHERE id = $1",
        )
        .bind(id)
        .fetch_one(&self.db)
        .await?;

        let template = PromptTemplateRecord {
            id: row.get("id"),
            name: row.get("name"),
            description: row.get("description"),
            template: row.get("template"),
            input_variables: row.get("input_variables"),
            provider: row.get("provider"),
            created_at: row.get("created_at"),
        };

        Ok(template)
    }

    /// Search prompt templates with pagination
    pub async fn search_prompt_templates(
        &self,
        name: Option<String>,
        _provider: Option<String>,
        page: i64,
        limit: i64,
    ) -> Result<Vec<PromptTemplateRecord>> {
        let offset = (page - 1) * limit;
        let rows = sqlx::query(
            "SELECT id, name, description, template, input_variables, provider, created_at
             FROM ai_prompt_templates
             WHERE ($1 IS NULL OR name LIKE $1)
             OFFSET $2 LIMIT $3",
        )
        .bind(name)
        .bind(offset)
        .bind(limit)
        .fetch_all(&self.db)
        .await?;

        let templates = rows
            .into_iter()
            .map(|row| PromptTemplateRecord {
                id: row.get("id"),
                name: row.get("name"),
                description: row.get("description"),
                template: row.get("template"),
                input_variables: row.get("input_variables"),
                provider: row.get("provider"),
                created_at: row.get("created_at"),
            })
            .collect::<Vec<_>>();

        Ok(templates)
    }

    /// Add prompt template
    pub async fn add_prompt_template(
        &self,
        name: &str,
        description: Option<String>,
        template: &str,
        input_variables: Option<String>,
    ) -> Result<i64> {
        let row = sqlx::query(
            "INSERT INTO ai_prompt_templates (name, description, template, input_variables)
             VALUES ($1, $2, $3, $4)
             RETURNING id",
        )
        .bind(name)
        .bind(description)
        .bind(template)
        .bind(input_variables)
        .fetch_one(&self.db)
        .await?;

        Ok(row.get("id"))
    }

    /// Update prompt template
    pub async fn update_prompt_template(
        &self,
        id: i64,
        name: Option<String>,
        description: Option<String>,
        template: Option<String>,
        input_variables: Option<String>,
    ) -> Result<i64> {
        let row = sqlx::query(
            "UPDATE ai_prompt_templates
             SET name = COALESCE($1, name),
                 description = COALESCE($2, description),
                 template = COALESCE($3, template),
                 input_variables = COALESCE($4, input_variables)
             WHERE id = $5
             RETURNING id",
        )
        .bind(name)
        .bind(description)
        .bind(template)
        .bind(input_variables)
        .bind(id)
        .fetch_one(&self.db)
        .await?;

        Ok(row.get("id"))
    }
}

File: backend/datasource/dbdao/ai_generation.go
package dbdao

import (
	"context"
	"errors"
	"time"

	"gorm.io/gorm"
)

type GeneratorModel struct {
	ID          int64     `gorm:"primarykey;column:id"`
	Name        string    `gorm:"column:name;type:varchar(255);not null"`
	Provider    string    `gorm:"column:provider;type:varchar(100);not null"`
	Model       string    `gorm:"column:model;type:varchar(100);not null"`
	APIKey      string    `gorm:"column:api_key;type:varchar(500)"`
	APIEndpoint string    `gorm:"column:api_endpoint;type:varchar(500)"`
	Enabled     bool      `gorm:"column:enabled;default:true"`
	CreatedAt   time.Time `gorm:"column:created_at;autoCreateTime:milli"`
}

func (GeneratorModel) TableName() string {
	return "ai_generators"
}

type Generator struct {
	ID          int64
	Name        string
	Provider    string
	Model       string
	APIKey      string
	APIEndpoint string
	Enabled     bool
	CreatedAt   time.Time
}

func (d *DB) GetGenerator(ctx context.Context, id int64) (*Generator, error) {
	var model GeneratorModel
	result := d.DB().WithContext(ctx).Where("id = ?", id).First(&model)
	if result.Error != nil {
		if result.Error == gorm.ErrRecordNotFound {
			return nil, errors.New("generator not found")
		}
		return nil, result.Error
	}

	return &Generator{
		ID:          model.ID,
		Name:        model.Name,
		Provider:    model.Provider,
		Model:       model.Model,
		APIKey:      model.APIKey,
		APIEndpoint: model.APIEndpoint,
		Enabled:     model.Enabled,
		CreatedAt:   model.CreatedAt,
	}, nil
}

func (d *DB) SearchGenerators(
	ctx context.Context,
	provider string, model string, enabled *bool,
	page, limit int64,
) ([]*Generator, int64, error) {
	var models []GeneratorModel
	q := d.DB().WithContext(ctx)

	if provider != "" {
		q = q.Where("provider = ?", provider)
	}
	if model != "" {
		q = q.Where("model = ?", model)
	}
	if enabled != nil {
		q = q.Where("enabled = ?", *enabled)
	}

	var total int64
	q.Model(&GeneratorModel{}).Count(&total)

	result := q.Offset(int((page - 1) * limit)).Limit(int(limit)).Find(&models)
	if result.Error != nil {
		return nil, 0, result.Error
	}

	gens := make([]*Generator, len(models))
	for i, m := range models {
		gens[i] = &Generator{
			ID:          m.ID,
			Name:        m.Name,
			Provider:    m.Provider,
			Model:       m.Model,
			APIKey:      m.APIKey,
			APIEndpoint: m.APIEndpoint,
			Enabled:     m.Enabled,
			CreatedAt:   m.CreatedAt,
		}
	}

	return gens, total, nil
}

func (d *DB) CreateGenerator(ctx context.Context, gen *Generator) (*Generator, error) {
	model := &GeneratorModel{
		Name:        gen.Name,
		Provider:    gen.Provider,
		Model:       gen.Model,
		APIKey:      gen.APIKey,
		APIEndpoint: gen.APIEndpoint,
		Enabled:     gen.Enabled,
	}

	result := d.DB().WithContext(ctx).Create(model)
	if result.Error != nil {
		return nil, result.Error
	}

	return &Generator{
		ID:          model.ID,
		Name:        model.Name,
		Provider:    model.Provider,
		Model:       model.Model,
		APIKey:      model.APIKey,
		APIEndpoint: model.APIEndpoint,
		Enabled:     model.Enabled,
		CreatedAt:   model.CreatedAt,
	}, nil
}

func (d *DB) UpdateGenerator(ctx context.Context, gen *Generator) (*Generator, error) {
	updates := map[string]interface{}{}

	if gen.Name != "" {
		updates["name"] = gen.Name
	}
	if gen.Provider != "" {
		updates["provider"] = gen.Provider
	}
	if gen.Model != "" {
		updates["model"] = gen.Model
	}
	if gen.APIKey != "" {
		updates["api_key"] = gen.APIKey
	}
	if gen.APIEndpoint != "" {
		updates["api_endpoint"] = gen.APIEndpoint
	}

	result := d.DB().WithContext(ctx).Model(&GeneratorModel{}).Where("id = ?", gen.ID).Updates(updates)
	if result.Error != nil {
		return nil, result.Error
	}

	if result.RowsAffected == 0 {
		return nil, errors.New("generator not found")
	}

	return d.GetGenerator(ctx, gen.ID)
}

type PromptTemplateModel struct {
	ID             int64     `gorm:"primarykey;column:id"`
	Name           string    `gorm:"column:name;type:varchar(255);not null"`
	Description    string    `gorm:"column:description;type:varchar(500)"`
	Template       string    `gorm:"column:template;type:longtext;not null"`
	InputVariables string    `gorm:"column:input_variables;type:json"`
	CreatedAt      time.Time `gorm:"column:created_at;autoCreateTime:milli"`
}

func (PromptTemplateModel) TableName() string {
	return "ai_prompt_templates"
}

type PromptTemplate struct {
	ID             int64
	Name           string
	Description    string
	Template       string
	InputVariables string
	CreatedAt      time.Time
}

func (d *DB) GetPromptTemplate(ctx context.Context, id int64) (*PromptTemplate, error) {
	var model PromptTemplateModel
	result := d.DB().WithContext(ctx).Where("id = ?", id).First(&model)
	if result.Error != nil {
		if result.Error == gorm.ErrRecordNotFound {
			return nil, errors.New("prompt template not found")
		}
		return nil, result.Error
	}

	return &PromptTemplate{
		ID:             model.ID,
		Name:           model.Name,
		Description:    model.Description,
		Template:       model.Template,
		InputVariables: model.InputVariables,
		CreatedAt:      model.CreatedAt,
	}, nil
}

func (d *DB) SearchPromptTemplates(
	ctx context.Context,
	name, provider string,
	page, limit int64,
) ([]*PromptTemplate, int64, error) {
	var models []PromptTemplateModel
	q := d.DB().WithContext(ctx)

	if name != "" {
		q = q.Where("name LIKE ?", "%"+name+"%")
	}

	var total int64
	q.Model(&PromptTemplateModel{}).Count(&total)

	result := q.Offset(int((page - 1) * limit)).Limit(int(limit)).Find(&models)
	if result.Error != nil {
		return nil, 0, result.Error
	}

	pts := make([]*PromptTemplate, len(models))
	for i, m := range models {
		pts[i] = &PromptTemplate{
			ID:             m.ID,
			Name:           m.Name,
			Description:    m.Description,
			Template:       m.Template,
			InputVariables: m.InputVariables,
			CreatedAt:      m.CreatedAt,
		}
	}

	return pts, total, nil
}

func (d *DB) CreatePromptTemplate(ctx context.Context, pt *PromptTemplate) (*PromptTemplate, error) {
	model := &PromptTemplateModel{
		Name:           pt.Name,
		Description:    pt.Description,
		Template:       pt.Template,
		InputVariables: pt.InputVariables,
	}

	result := d.DB().WithContext(ctx).Create(model)
	if result.Error != nil {
		return nil, result.Error
	}

	return &PromptTemplate{
		ID:             model.ID,
		Name:           model.Name,
		Description:    model.Description,
		Template:       model.Template,
		InputVariables: model.InputVariables,
		CreatedAt:      model.CreatedAt,
	}, nil
}

func (d *DB) UpdatePromptTemplate(ctx context.Context, pt *PromptTemplate) (*PromptTemplate, error) {
	updates := map[string]interface{}{}

	if pt.Name != "" {
		updates["name"] = pt.Name
	}
	if pt.Description != "" {
		updates["description"] = pt.Description
	}
	if pt.Template != "" {
		updates["template"] = pt.Template
	}
	if pt.InputVariables != "" {
		updates["input_variables"] = pt.InputVariables
	}

	result := d.DB().WithContext(ctx).Model(&PromptTemplateModel{}).Where("id = ?", pt.ID).Updates(updates)
	if result.Error != nil {
		return nil, result.Error
	}

	if result.RowsAffected == 0 {
		return nil, errors.New("prompt template not found")
	}

	return d.GetPromptTemplate(ctx, pt.ID)
}

type GenerateTaskModel struct {
	ID          int64      `gorm:"primarykey;column:id"`
	GeneratorID int64      `gorm:"column:generator_id;index"`
	Input       string     `gorm:"column:input;type:longtext"`
	Output      string     `gorm:"column:output;type:longtext"`
	Status      string     `gorm:"column:status;type:varchar(50)"`
	Error       string     `gorm:"column:error;type:text"`
	CreatedAt   time.Time  `gorm:"column:created_at;autoCreateTime:milli"`
	StartedAt   *time.Time `gorm:"column:started_at"`
	CompletedAt *time.Time `gorm:"column:completed_at"`
}

func (GenerateTaskModel) TableName() string {
	return "ai_generate_tasks"
}

type GenerateTask struct {
	ID          int64
	GeneratorID int64
	Input       string
	Output      string
	Status      string
	Error       string
	CreatedAt   time.Time
	StartedAt   *time.Time
	CompletedAt *time.Time
}

func (d *DB) CreateGenerateTask(ctx context.Context, task *GenerateTask) (*GenerateTask, error) {
	model := &GenerateTaskModel{
		GeneratorID: task.GeneratorID,
		Input:       task.Input,
		Status:      task.Status,
	}

	result := d.DB().WithContext(ctx).Create(model)
	if result.Error != nil {
		return nil, result.Error
	}

	return &GenerateTask{
		ID:          model.ID,
		GeneratorID: model.GeneratorID,
		Input:       model.Input,
		Status:      model.Status,
		CreatedAt:   model.CreatedAt,
	}, nil
}
